File: src/lib/upHeroRng.ts
/**
 * Phase 14 code-review High #7 — Up Hero 전투는 18+ 개 Math.random() 사이트에
 *   의존하고 있어 (a) 재현성 있는 시뮬레이션 / regression test 가 불가능, (b) 주간
 *   리더보드에서 같은 세션을 공정하게 비교 불가. seedable PRNG 를 도입해 선택적
 *   결정론 확보. 기본 동작은 Math.random 으로 유지 (프로덕션 체감 변화 없음).
 *
 *   테스트 / dev 경로에서 `setRngSeed(seed)` 호출 시 mulberry32 기반 deterministic
 *   sequence 로 전환. 세션 끝나거나 명시 `resetRng()` 시 Math.random 복귀.
 *
 *   TODO(Phase 15): session.rngSeed 필드 persist → cloud replay 지원.
 */

type RngFn = () => number;

let activeRng: RngFn | null = null;

/**
 * mulberry32 — 32-bit state, period 2^32, 균등 분포 정합성 충분.
 *   seed 0 은 degenerate (항상 동일 값) 이라 내부에서 nonzero 로 정규화.
 */
export function createRng(seed: number): RngFn {
  let s = (seed | 0) || 0x9e3779b9; // 0 → golden ratio 로 정규화.
  return () => {
    s = (s + 0x6d2b79f5) | 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/**
 * 전역 RNG seed 설정. 이 호출 이후 `rng()` 는 deterministic sequence 반환.
 *   `null` / `undefined` seed 는 Math.random 로 복귀시킴 (resetRng 과 동등).
 */
export function setRngSeed(seed: number | null | undefined): void {
  activeRng = seed == null ? null : createRng(seed);
}

export function resetRng(): void {
  activeRng = null;
}

/**
 * 현재 활성 RNG 로부터 [0, 1) 숫자 추출. 미설정 시 Math.random 로 위임.
 *   upHeroCombat 의 모든 확률 분기는 이 함수로 일원화돼 결정론 전환이 가능.
 */
export function rng(): number {
  return activeRng ? activeRng() : Math.random();
}

File: upnext-ios/UpNext/UpNext/Models/UpHeroRNG.swift
//
//  UpHeroRNG.swift
//  UpNext 로직 — 결정론적 PRNG (mulberry32).
//
//  웹 src/lib/upHeroRng.ts를 비트 단위로 1:1 포팅.
//  Up Hero 전투의 모든 확률 분기가 이 RNG를 통과 → 같은 seed면 같은 전투 재현.
//  Phase 2.5 동치성 검증의 핵심 (같은 seed → 웹과 Swift가 동일 수열).
//
//  JS Math.imul / >>> / |0 을 Swift UInt32 wrapping 연산(&+, &*, >>)으로 정확 환산:
//   - JS 32-bit 정수 연산 ≡ Swift UInt32 wrapping
//   - JS `t >>> n` (unsigned shift) ≡ Swift UInt32 `t >> n`
//   - JS `Math.imul(a,b)` (32-bit 곱) ≡ Swift `a &* b`
//
//  Phase 2.3 (결정론적 알고리즘 Swift 포팅) 산출물.
//

import Foundation

/// 게임 로직이 쓰는 난수 소스 — [0, 1) 실수를 제공. 웹 `rng()` 대응.
protocol RandomSource {
    /// [0, 1) 범위 균등 분포 난수. 웹 rng()와 동일 시맨틱.
    mutating func unit() -> Double
}

/// mulberry32 결정론 PRNG. 웹 `createRng(seed)`와 비트 단위 동일.
/// 32-bit state, 주기 2^32. seed 0은 degenerate라 golden ratio로 정규화.
struct Mulberry32: RandomSource {
    private var state: UInt32

    init(seed: Int) {
        // 웹 `(seed | 0) || 0x9e3779b9` — 32-bit 절단 후 0이면 golden ratio.
        let s = UInt32(truncatingIfNeeded: seed)
        state = (s == 0) ? 0x9e3779b9 : s
    }

    mutating func unit() -> Double {
        // 웹 createRng 클로저 본문과 1:1.
        state = state &+ 0x6d2b79f5
        var t = state
        t = (t ^ (t >> 15)) &* (t | 1)
        t = t ^ (t &+ ((t ^ (t >> 7)) &* (t | 61)))
        return Double(t ^ (t >> 14)) / 4294967296.0
    }
}

/// 시스템 난수 소스 — 프로덕션 기본값. 웹 `Math.random` 경로 대응.
struct SystemRandom: RandomSource {
    mutating func unit() -> Double {
        Double.random(in: 0..<1)
    }
}

// MARK: - RandomSource 공용 헬퍼

extension RandomSource {
    /// [0, n) 정수. 웹 `Math.floor(rng() * n)` 패턴 대응.
    mutating func int(below n: Int) -> Int {
        guard n > 0 else { return 0 }
        return min(n - 1, Int(unit() * Double(n)))
    }

    /// 확률 p로 true. 웹 `rng() < p` 패턴 대응.
    mutating func chance(_ p: Double) -> Bool {
        unit() < p
    }

    /// 배열에서 균등 랜덤 1개. 비었으면 nil.
    mutating func pick<T>(_ array: [T]) -> T? {
        guard !array.isEmpty else { return nil }
        return array[int(below: array.count)]
    }
}
